### ISSUE TITLE
[BLOCKER] Desktop Runtime Error: "no such file or directory client/server/bundler/assets-fallback.json"

### ISSUE THREAD

**Issue by nsakaimbo** [Automattic/wp-calypso#42163]
### Description

Looks like a recent change to Calypso is driving a new runtime error in wp-desktop: :`no such file or directory client/server/bundler/assets-fallback.json`

As far as I can tell, what's really strange about this error is we're only getting it in CI - it seems the app can still build just fine locally.

I see in Git history there were some recent changes made in 4458c1b2d19e0dd652d615fded9a25c3561371c8 that look related, although maybe a red herring. @sgomes looks like you worked on this specific change, are you able to help take a look? Happy to pair on this as this is seems really bizarre (as it's affecting branches that were building successfully before. Maybe something in the desktop webpack config needs to be updated?).

**Context**:

We [recently added desktop end-to-end testing](https://github.com/Automattic/wp-desktop/pull/870) to the Calypso pipeline (on 5/5), but looks like there were some webpack changes merged prior to then.

**Comment by nsakaimbo** [comment]
**Update**: Confirmed the root cause is the `build-languages-if-enabled` script (which is a relatively recent addition). The script results in a corrupt Calypso build for the desktop app. The desktop app is _not_ compiled as a chunked build, and this might be the primary conflict.

Confirmed that building with this this script disabled (as with [this](https://github.com/Automattic/wp-calypso/compare/try-desktop-remove-build-languages-if-enabled) test branch) allows the desktop app to continue building as normal.

Build job: [link](https://app.circleci.com/pipelines/github/Automattic/wp-desktop/13668/workflows/4aab50ea-3363-4995-aef9-4fab0586019c)

cc: @yuliyan @akirk

**Comment by yuliyan** [comment]
@nsakaimbo #42204 that will completely exclude `build-languages-if-enabled` from desktop builds and only execute it with `build-client-if-prod` script.

However, `build-languages-if-enabled` is supposed to check if either `BUILD_TRANSLATION_CHUNKS=true` or `ENABLE_FEATURES=use-translation-chunks` env variables are set. If you are not setting either of those, it should skip the actual process of building language files.

**Comment by nsakaimbo** [comment]
Thanks for your help, @yuliyan. 

After more experimentation/debugging, we were able to determine this script isn't the source of the issue. It seems to be driven by the minification step, which doesn't play nice with the desktop's non-chunked build (see update below). However, if this script isn't necessary for the desktop build, we might as well exclude it.

**Comment by nsakaimbo** [comment]
**Update**:

Seems dealing with an out-of-memory error resulting in the Calypso build being killed. TL;DR:

- Calypso is failing to build for wp-desktop, for some reason the Calypso build process is killed midway
- There is a bug in the CI script that still marks the Calypso build as successful and "saves it to the cache" (corrupted) anyway
- On downstream jobs and subsequent builds, the corrupt Calypso cache is restored and is used to build a corrupt desktop executable
- The reason we're unable to reproduce this locally is because on a local machine we don't have the same `max_old_space_size` constraints as in CI (i.e. for the desktop it has to be strictly within the machine's RAM of 4GB, not 8GB). The Calypso process is killed when it hits the `max_old_space_size` ceiling.

Further investigation revealed that [turning off minification](https://github.com/Automattic/wp-calypso/compare/try-disable-minimize) for the desktop build allows things to continue to build. The resulting built source file `build/desktop.js` is still more than a reasonable size (~6MB) for a desktop application, and smoke-testing indicates the app seems to continue to work as expected. _Note_: I did try to do this purely from the desktop by setting MINIFY_JS=true, but unfortunately that resulted in an error (look at build-source step of `https://app.circleci.com/pipelines/github/Automattic/wp-desktop/13750/workflows/e5383f4a-5326-499b-9f13-d45d54c3adb5/jobs/61413`)

We'd like to revisit chunking at a later date, but if we can turn off minimization for the desktop build in `packages/calypso-build/webpack.config.js` that would unblock us for the time being.

**Comment by nsakaimbo** [comment]
Another update:

Turns out this exact error has been reported specifically for Terser within the context of Circle CI: ([ref](https://github.com/facebook/create-react-app/issues/8320)): `CircleCI + Terser Error: Call retries were exceeded`.

A couple of highlights:

* In Circle CI, `os.cpus().length` [returns many more CPUs](https://github.com/webpack-contrib/terser-webpack-plugin/issues/143#issuecomment-573954013) than are allocated to your container
* Setting the `parallel` option of TerserWebpackPlugin is strongly recommended
* We can probably further by getting rid of unnecessary optimizations when building for desktop, e.g. mangling and Safari in `packages/calypso-build/webpack.config.js`

I tried:

* Updating terser to latest: `https://app.circleci.com/pipelines/github/Automattic/wp-desktop/13763/workflows/246d6284-2c2b-4b99-87c9-1de1efe14cff`

and 

* turning off safari and mangling optimizations in packages webpack config: `https://app.circleci.com/pipelines/github/Automattic/wp-desktop/13765/workflows/0ddf7eec-b26e-4c5f-84dd-a28cf4efd1cf`

Both of these seemed to run successfully. Not sure if this is indicative of a potential solution or another red herring.
